specs/core/0_beacon-chain.md

```diff
@@ -1771,80 +1774,58 @@ def maybe_reset_eth1_period(state: BeaconState) -> None:
     eligible_validators = [
         index for index, validator in enumerate(state.validator_registry)
         if (
-            is_active_validator(validator, current_epoch) or
-            (validator.slashed and current_epoch < validator.withdrawable_epoch)
+            is_active_validator(validator, previous_epoch) or
+            (validator.slashed and previous_epoch < validator.withdrawable_epoch)
         )
     ]
-    # Process rewards or penalties for all validators
+    rewards = [0 for index in range(len(state.validator_registry))]
+    penalties = [0 for index in range(len(state.validator_registry))]
     for index in eligible_validators:
-        base_reward = get_base_reward(state, index)
-        # Expected FFG source
+        base_reward = get_base_reward(state, get_previous_epoch_total_balance(state), index)
+
+        # Micro-incentives for matching FFG source, matching FFG target, and matching head 
+        for attestations in (
+            state.previous_epoch_attestations,                      # Matching FFG source
+            get_previous_epoch_matching_target_attestations(state), # Matching FFG target
+            get_previous_epoch_matching_head_attestations(state),   # Matching head
+        ):
+            if index in get_unslashed_attesting_indices(state, attestations):
+                rewards[index] += base_reward * get_attesting_balance(state, attestations) // get_previous_epoch_total_balance(state)
+            else:
+                penalties[index] += base_reward
+
+        # Inclusion delay micro-penalty
         if index in get_unslashed_attesting_indices(state, state.previous_epoch_attestations):
-            rewards[index] += base_reward * total_attesting_balance // total_balance
-            # Inclusion speed bonus
             earliest_attestation = get_earliest_attestation(state, state.previous_epoch_attestations, index)
             inclusion_delay = earliest_attestation.inclusion_slot - earliest_attestation.data.slot
-            rewards[index] += base_reward * MIN_ATTESTATION_INCLUSION_DELAY // inclusion_delay
-        else:
-            penalties[index] += base_reward
-        # Expected FFG target
-        if index in get_unslashed_attesting_indices(state, boundary_attestations):
-            rewards[index] += base_reward * boundary_attesting_balance // total_balance
-        else:
-            penalties[index] += get_inactivity_penalty(state, index, epochs_since_finality)
-        # Expected head
-        if index in get_unslashed_attesting_indices(state, matching_head_attestations):
-            rewards[index] += base_reward * matching_head_balance // total_balance
-        else:
-            penalties[index] += base_reward
-        # Take away max rewards if we're not finalizing
-        if epochs_since_finality > 4:
-            penalties[index] += base_reward * 4
+            penalties[index] += base_reward * (1 - MIN_ATTESTATION_INCLUSION_DELAY // inclusion_delay)
```

Comment:
```suggestion
            penalties[index] += base_reward if inclusion_delay >= MIN_ATTESTATION_INCLUSION_DELAY else 0
```
I would prefer using logic instead of arithmetics for readability

Reply:
Oh that's not what I intended. We always have `inclusion >= MIN_ATTESTATION_INCLUSION_DELAY`.

I probably want `base_reward * (inclusion_delay - MIN_ATTESTATION_INCLUSION_DELAY) // inclusion_delay`.